larpandoracontent/LArTwoDReco/HitWidthClusterMergingAlgorithm.cc

```diff
@@ -0,0 +1,503 @@
+{
+
+
+HitWidthClusterMergingAlgorithm::HitWidthClusterMergingAlgorithm() :
+  m_clusterListName(),
+  m_maxConstituentHitWidth(0.5),
+  m_hitWidthScalingFactor(1.0),
+  m_useSlidingLinearFit(false),
+  m_layerFitHalfWindow(20),
+  m_fittingWeight(10),
+  m_minClusterWeight(0.5),      
+  m_maxXMergeDistance(5),     
+  m_maxZMergeDistance(2),     
+  m_maxMergeCosOpeningAngle(0.97),
+  m_maxDirectionDeviationCosAngle(0.90)
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void HitWidthClusterMergingAlgorithm::GetListOfCleanClusters(const ClusterList *const pClusterList, ClusterVector &clusterVector) const
+{
+ 
+    LArHitWidthHelper::ClusterToParametersMapStore* pClusterToParametersMapStore = LArHitWidthHelper::ClusterToParametersMapStore::Instance();
+    LArHitWidthHelper::ClusterToParametersMap* pClusterToParametersMap = pClusterToParametersMapStore->GetMap();
+
+    // Clear map if already full i.e. from other view clustering
+    if(!pClusterToParametersMap->empty())
+        pClusterToParametersMap->clear();
+
+    for (const Cluster *const pCluster : *pClusterList)
+    {
+        // The original cluster weight, with no hit scaling or hit padding
+        if(LArHitWidthHelper::GetOriginalTotalClusterWeight(pCluster) < m_minClusterWeight)
+            continue;
+
+        pClusterToParametersMap->insert(std::pair<const Cluster*, LArHitWidthHelper::ClusterParameters>(pCluster, LArHitWidthHelper::ClusterParameters(pCluster, m_maxConstituentHitWidth, m_useSlidingLinearFit, m_hitWidthScalingFactor)));
+
+        clusterVector.push_back(pCluster);
+    }
+
+    //ORDER BY MAX EXTREMAL X COORDINATE
+    std::sort(clusterVector.begin(), clusterVector.end(), LArHitWidthHelper::SortByHigherXExtrema);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void HitWidthClusterMergingAlgorithm::PopulateClusterAssociationMap(const ClusterVector &clusterVector, ClusterAssociationMap &clusterAssociationMap) const
+{
+
+    LArHitWidthHelper::ClusterToParametersMapStore* pClusterToParametersMapStore = LArHitWidthHelper::ClusterToParametersMapStore::Instance();
+    LArHitWidthHelper::ClusterToParametersMap* pClusterToParametersMap = pClusterToParametersMapStore->GetMap();
+
+    if(pClusterToParametersMap->empty())
+        throw StatusCodeException(STATUS_CODE_NOT_INITIALIZED);
+
+    // ATTN this method assumes that clusters have been sorted by extremal x position (low x -> high x)
+    for (ClusterVector::const_iterator iterCurrentCluster = clusterVector.begin(); iterCurrentCluster != clusterVector.end(); ++iterCurrentCluster)
+    {
+        const Cluster *const pCurrentCluster = *iterCurrentCluster;
+        const LArHitWidthHelper::ClusterToParametersMap::const_iterator currentParametersIter = pClusterToParametersMap->find(pCurrentCluster);
```

Comment:
happy to make this change - I haven't seen auto used often in Pandora, is this because the majority of the code was written before the introduction of the keyword and hence should now be used more often?

Reply:
Probably not - auto has been around for a while, so it probably has more to do with wanting to be explicit about the objects you are working with (John can probably comment here). However, when it comes to things like iterators, where you're only really using them as a useful looping construct and they usually have horrendous type names, I think retaining the exact type information is much less important and so can be conveniently hidden away by auto, particularly where you assign what the iterator is pointing to, to a named type, as you do here with `pCurrentCluster` for example. The main thing to avoid is a proliferation of autos where it becomes difficult to understand what anything is.